crates/node/src/providers/robust_ecdsa/presign.rs

```diff
@@ -0,0 +1,293 @@
+    /// generation (i.e. other participants of the computations this function initiates),
+    /// so that needs to be separately handled.
+    pub(super) async fn run_background_presignature_generation(
+        client: Arc<MeshNetworkClient>,
+        mpc_config: Arc<MpcConfig>,
+        config: Arc<PresignatureConfig>,
+        domain_id: DomainId,
+        presignature_store: Arc<PresignatureStorage>,
+        keygen_out: KeygenOutput,
+    ) -> anyhow::Result<()> {
+        let in_flight_generations = InFlightGenerationTracker::new();
+        let progress_tracker = Arc::new(PresignatureGenerationProgressTracker {
+            desired_presignatures_to_buffer: config.desired_presignatures_to_buffer,
+            presignature_store: presignature_store.clone(),
+            in_flight_generations: in_flight_generations.num_in_flight_atomic(),
+        });
+        let parallelism_limiter = Arc::new(tokio::sync::Semaphore::new(config.concurrency));
+        let mut tasks = AutoAbortTaskCollection::new();
+
+        let running_participants: Vec<_> = mpc_config
+            .participants
+            .participants
+            .iter()
+            .map(|p| p.id)
+            .collect();
+        let threshold = mpc_config.participants.threshold as usize;
+
+        loop {
+            progress_tracker.update_progress();
+            metrics::MPC_OWNED_NUM_PRESIGNATURES_ONLINE
+                .set(presignature_store.num_owned_ready() as i64);
+            metrics::MPC_OWNED_NUM_PRESIGNATURES_WITH_OFFLINE_PARTICIPANT
+                .set(presignature_store.num_owned_offline() as i64);
+            let my_presignatures_count: usize = presignature_store.num_owned();
+            metrics::MPC_OWNED_NUM_PRESIGNATURES_AVAILABLE.set(my_presignatures_count as i64);
+            let should_generate = my_presignatures_count + in_flight_generations.num_in_flight()
+                < config.desired_presignatures_to_buffer;
+            if should_generate
+                // There's no point to issue way too many in-flight computations, as they
+                // will just be limited by the concurrency anyway.
+                && in_flight_generations.num_in_flight()
+                < config.concurrency * 2
+            {
+                let id = presignature_store.generate_and_reserve_id();
+                let participants = match client.select_random_active_participants_including_me(
+                    threshold,
+                    &running_participants,
+                ) {
+                    Ok(participants) => participants,
+                    Err(e) => {
+                        tracing::warn!(
+                            "Can't choose active participants for a triple: {}. Sleeping.",
+                            e
+                        );
+                        // that should not happen often, so sleeping here is okay
+                        tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
we should certainly have an issue for removing these sleep ops, not doing so just in case something actually breaks

Reply:
Ouch yeah this is bad. Sleeping here is not okay, it makes anything touching these code paths untestable.